keps/sig-api-machinery/20190422-cloud-controller-manager-migration.md

```diff
@@ -0,0 +1,281 @@
+to their out-of-tree equivalents (via cloud-controller-manager). The migration process laid out here should be reusable by other components in the future if desired.
+
+## Motivation
+
+SIG Cloud Provider is in the process of migrating the cloud specific code from the core Kubernetes tree to external packages
+and removing them from the kube-controller-manager, where they are today embedded. Once the extraction has been completed, existing users
+running older versions of Kubernetes need a process to migrate their existing clusters to use the new cloud-controller-manager component
+with minimal risk. This KEP proposes a mechanism in which HA clusters can safely migrate “cloud specific” controllers between the
+kube-controller-manager and the cloud-controller-manager via a shared resource lock between the two components. The pattern proposed
+in this KEP should be reusable by other components in the future if desired.
+
+### Goals
+
+* Define a clear migration process for existing clusters to use the cloud-controller-manager instead of the kube-controller-manager for cluster integrations with any cloud provider.
+* The migration process should work for single node control planes and HA control planes with minimal risk to users.
+
+### Non-Goals
+
+* Removing cloud provider code from the core Kubernetes tree, this effort is separate and is covered in [KEP-removing-in-tree-providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md)
+* Improving the scalability of controllers by running controllers across multiple components (with or without leader election).
+* Migrating cloud-based volume plugins to CSI. This is a separate effort led by SIG Storage. See [this proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md) for more details.
+
+## Proposal
+
+Today, the kube-controller-manager and cloud-controller-manager run independent of each other.
+This means that both the kube-controller-manager or the cloud-controller-manager can run the
+cloud specific control loops for a given cluster. For a highly available control plane to migrate
+from running only the kube-controller-manager to running both the kube-controller-manager and
+the cloud-controller-manager requires that only one process in the control plane is running
+the cloud specific controllers. This becomes non-trivial when introducing the cloud-controller-manager
+that runs overlapping controllers with the kube-controller-manager.
+
+For environments that can tolerate downtime, it will be recommended to shutdown the control plane,
+reconfigure components to use the cloud-controller-manager, and then bring the control plane back up.
+This ensures that only 1 component can be running the set of cloud controllers. For environments
```

Comment:
"This allows the existing leader election mechanism to guarantee that at most one instance of each controller is running."